README

m2vmp2cut 0.87-wip

m2vmp2cut is command-line driven frame accurate (currently PAL) mpeg2
video (m2v file) with accompanied mp2 audio (mp2 file) cutter.
A GUI tool is used for selecting the cutpoints.

Frame accuracy is achieved by re-encoding video around cutpoints -- or
sacrifice a bit of selection choices to get output video without any
re-encoding required.

Audio is cut from separate mp2 files at positions that keep a/v sync as
good as possible (maximum sync difference is around 10-15 milliseconds).

This program is licensed under the GPL v2. See file COPYING for details.


BASIC WORKFLOW:

Note that 'm2vmp2cut' used to have pretty good help built-in. Executing
'm2vmp2cut' or 'm2vmp2cut . help' will get to it...

First run

$ m2vmp2cut <mpeg2 file> demux <more mpeg2 files or other ProjectX opts>

to demultiplex the source mpeg2.

Now, if you suspect there may be broken subtitle image files (the output
of demux will hint you for this), run

$ m2vmp2cut <mpeg2 file/demux outputdir> pp fixbmps

to verify (and fix) the files.

Next execute GUI for cutpoint hunt:

$ m2vmp2cut <mpeg2 file/demux outputdir> select

<demux outputdir> is derived from the original mpeg2 file by removing latest
suffix after dot (.) and replacing that with '.d' (now it also works without
the trailing 'd' (for user convenience, common prefix for tab completion)).

Select will first show gui window where you can choose what audio and subtitle
files to mux after select. Note that currently these apply to icut & imkvcut
only (the old 'cut' uses only the first audio file chosen by this tool).
When this is done, the graphical content selection tool will start.

When selecting cutpoints it is always good habit to test how the end of
previous part and start of next part will "glue" together. Just press 't'
when at either of these cutpoints to get a test play around these cutpoints
to happen. Also remember 'h' command to see keybindings (and forget mouse
for the time being).

After selection of the cutpoints are done, there are a few options to
continue: The old m2vmp2cut 'cut' command will create new mpeg2 file from
the selected contents, re-encoding some video frames to achieve frame
accuracy. Audio frames don't require any re-encoding.

But latest development has gone to 'pp icut' and 'pp imkvcut' commands.
These 2 require that in each part the start frame is 'I' frame and end
frame is either 'I' frame or 'P' frame. From these selections new video
track can be constructed without re-encoding anything. The imkvcut command
uses all of the audio and subtitle options in the pre-select gui. icut
can only use audio and language info is not used there.

Last, there are other tools in 'pp' which can create content (and in the
future this may even improve) and also m2vmp2cut 'getyuv' and 'getmp2'
may be used in other tools...


REQUIREMENTS:

gtk2            -- gui library used by m2vcut
libmpeg2        -- for decoding mpeg2 pictures
Mjpegtools      -- for mpeg2enc and mplex tools
Mkvtoolnix      -- for mkvmerge
feh             -- for viewing subtitle pictures
mplayer         -- for viewing test cuts and playing audio options
ProjecX (project-x.sourceforge.net) -- to demux source mpeg files
Java runtime    -- to run ProjectX (many non-sun vm:s work also)

Perl            -- m2vmp2cut.pl (run frontend) is written in perl
Python          -- some m2vmp2cut utilities are written in python
"Modern" Shell  -- /bin/sh which has modern shell capabilities
                   (not just Bourne) (this is usually the case)

'C' compiler    -- there are some C utilities that needs to be compiled
Java compiler   -- if you need to compile ProjectX

(C quoted to avoid file(1) recognize this file as 'FORTRAN program,' ;)

Suggested:

Transcode       -- for requantize option (e.g. to fit large image on dvd)


(Other than ProjecX) Demux Tools (: (not used here (anymore))
mpgtx
replex
tcextract
ffmpeg
mplayer (dumpaudio/dumpvideo?)


USAGE:

Run 'm2vmp2cut . help' to get more usage information.


CONTRIBUTORS:

Göran Uddeborg
Daniel Pielmeier


WHY LIKE THIS:

Cutting demuxed files instead of mpeg2 file that has both audio and video
is done since:

1) Handling demuxed video file is easier; no messing with PTS, DTS and SCR
   and so on (I saved development time).

2) I can't compete with such a good tool as ProjectX to get stuff out
   of mpeg2 files.

3) Cutting separate audio and video files yields much better results cutting
   audio and video on right places (near-perfect A/V sync if original has
   so), for example using 48KHz 224 kbit mp2 files one 672 byte audio frame
   contains 24ms of audio -- so A/V sync is at max 12ms off. If doing
   'hard-cuts' in (multiplexed) mpeg2 files 1) audio packets contains like
   2.6 audio frames and 2) audio packets might be far off in the file
   the video packets with time close to each other...


Tools that helped to find out required information (Thanks to the authors!):

gopchop mpegcat tool
bbtool17-2-linux
dvb-mpegtools analyze
BDSup2Sub
google searches: `mpeghdrs.html' (video) and `mpeghdr.htm' (audio)
google search: `python_mpeg'


NEWS
m2vmp2cut 0.87-wip (unreleased) (Cutting Iron)
==============================================

(2015-02-19)

* m2vcut-gui: When selecting "end" cutpoint of a part the selection is
   inclusive. This makes selecting I or P frame for that more explicit.

* assel-gui: New program to choose audio and subtitle streams; run before
   m2vcut-gui when 'select' command is used.

* contrib has been renamed to pp; the most advanced tools (icut & imkvcut)
   are there. E.g. pp imkvcut can create .mkv file containing one video
   track accompanied with multible audio and subtitle (of pictures) tracks.

* m2vcut-gui: Changed pgup/pgdown keybindings to do 2 min jumps, but after
   first direction change, halve jump in each step.

* m2vcut-gui: Added 6 second jumps to shift+left/right arrow and 1 minute
   jumps to shift+up/down arrow keys.

* m2vcut-gui: The space showing audio levels of current frame now has either
   greenish(frame in) or reddish(frame out) background.

* pp/icut: Support for cutting without re-encoding anything added to
   various places in source code. Sections starting in I-frame and ending
   in I/P-frame can be cut without re-encoding.

* pp/imkvcut: Like icut above, additionally subtitle tracks can be "merged"
   to final file. Also language codes (set by assel-gui) are effective here.

* Some general code improvements and minor behaviour changes done.
* Fine-tuned some build system details.

* ... and probably many things not known until git logs and diff to 0.86
  examimed...

m2vmp2cut 0.86 (2012-10-18)
===========================

* Improved build system with user-definable optimization flags
  and some installation variable by importing and fitting patches
  from Göran Uddeborg.

m2vmp2cut 0.84 (2012-10-14)
===========================

* Added contrib/ffgetyuv.pl which uses ffmpeg to decode m2v input.
* Contrib scripts use ffgetyuv.pl for some m2v error concealment.
* Hid deprecated m2vmp2cut lvev6frames feature.
* Removed obsoleted 'a' (show agraph) feature from select GUI.
* Some small usage/doc improvements.

m2vmp2cut 0.82 (2012-09-30)
===========================

* m2vmp2cut cli now requires source file/dir as first arg (consistently).
* Approximate audio levels visualized in m2vcut.
* Added contrib/webm.sh script.
* Demux output files now written to subdirectory in current dir.
* General code improvemenet with no functional change.

m2vmp2cut 0.78 (2009-11-20) (Lost and Found)
============================================

* If 'projectx' command is found (in PATH), that is used.
* Demux can demux multiple files.
* Bug fix when last chars in file were \000\000\001.

m2vmp2cut 0.76 (2008-09-09)
===========================

* Fixed seeking to offset zero (-sb 0) in m2vtoyuv command.

m2vmp2cut 0.74 (2008-08-18)
===========================

* Fixed largefile support in fileparts command.
* Added getyuv.pl and getmp2.sh -- and related commands to m2vmp2cut;
  these outputs selected video and audio parts from the source.
* Added "contrib" interface.
* Added first "contrib" script, mpeg2 re-encoding.
* Tuned file/directory argument position on m2vmp2cut command (again).

m2vmp2cut 0.72 (2008-07-30)
===========================

* ANNOUNCE informs this is stable version.
* buildlibmpeg-051.sh accepts libmpeg-0.5.1.tar.gz with absolute path.
* Minor tunes, simplification etc.

m2vmp2cut 0.68 (2008-07-27) (Not even far!)
===========================================

* Binaries wavgraph and warpxpointer are installed.
* Audio sync fixed in cutpoint test functionality.
* ProjectX is linked for use during 'make install'.

0.67 (Success is not an option)
'''''''''''''''''''''''''''''''
* Audio visualizer around current position in m2vcut-gui, to find
  (silent) spot for cutpoint.
* Test speedups, now audio is pre-scanned too.
* Some usage tunes, fixes and enchangements.
* Bug fixes in almost every area.

0.65 (Faux pas)
'''''''''''''''
* Target size estimation.
* Keyboard control adjustments in gui.
* Bug fixes.

0.63 (Epic Fail)
''''''''''''''''
* Added graphical gui client for cutpoint selection.
* Added m2vtoyuv for frame decoding -- in place of transcode(1)
* Now gops are closed just after encoded frames.
* Added (cli) frontend to do run demuxer, cutter and m2vmp2cut.pl
*  Lots of other tunes

m2vmp2cut 0.62 (2008-07-27) (Gender-neutral pronouns)
=====================================================

* Makefile tail commands fixed to have -n before +<num>

m2vmp2cut 0.58 (2005-10-23) (Wrecking Stars)
============================================

* Timecodes in GOP's "fixed".
* Audio scanning implemented in C for speed...
  ... and if audio bit/samplerates change in output that will be an
  error.
* Source video format and intra_dc_precision value are used in
  video encoder (Still some settings that cannot be defined in
  encoder left, though).
* Some (a little less significant) improvements

m2vmp2cut 0.56 (2005-03-04) (Fifty-Sixty)
=========================================

* New option: --aonly (cut audio only (video to be re-encoded or so)).
* New option: --evbr (other way to prepare video for re-encoding...).
* Added "encoded frame count" checks.
* Uses lvemedia.lst in addition to lveedit.lst for finding cutpoints
  (to go around (IMHO) lve bug when dragging frame 0).
* Examples tuning/fixing.
* Added some lve patches to ./misc
* Internal workings changes.
* Bug fixes.

m2vmp2cut 0.54 (2004-11-03) (Constit on monet)
==============================================

* --sync option bugfix: negative numbers were taken as positive.
* --test option bugfix: was no-op in 0.42
* New options, --last, --requant, -. And --options, --examples.
* More embedded documentation.
* Added `misc' dir containing external stuff.
* Internal workings changes.

m2vmp2cut 0.42 (2004-10-21) (Infinite Improbabilities)
======================================================

* First release.


src/zzob.c
#if 0 /*
 TRG=`basename $0 .c`; rm -f "$TRG" #"${TRG}_defs.hg"
 #trap 'rm -f "${TRG}_defs.hg"' 0; perl -x "$0" "$0" > "${TRG}_defs.hg"
 WARN="-Wall -Wstrict-prototypes -pedantic -Wno-long-long"
 WARN="$WARN -Wcast-align -Wpointer-arith " # -Wfloat-equal #-Werror
 WARN="$WARN -W -Wwrite-strings -Wcast-qual -Wshadow" # -Wconversion
 FLAGS='-O2 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64' # -pg
 lastrun() { echo "$@"; exec "$@"; exit $?; }
 lastrun ${CC:-gcc} $WARN "$@" -o "$TRG" "$0" $FLAGS -DCDATE="\"`date`\""
 #*/
#endif
/*
 * zzob.c
 *
 * Author: Tomi Ollila too ät iki fi
 *
 *	Copyright (c) 2004 Tomi Ollila
 *	    All rights reserved
 *
 * Created: Thu Sep 23 19:28:18 EEST 2004 too
 * Last modified: Thu 25 Oct 2012 21:45:18 EEST too
 *
 * This program is licensed under the GPL v2. See file COPYING for details.
 */


#include <stdio.h> /* for NULL */
#include <string.h> /* for memmove() */
#include <unistd.h> /* for read() */

#define DBGS 0
#include "zzob.h"

#if 0
void __cyg_profile_func_enter(void *this_fn, void *call_site)
  __attribute ((no_instrument_function));
void __cyg_profile_func_enter(void *this_fn, void *call_site)
{
  write(1, "enter\n", 6);
}
void __cyg_profile_func_exit(void *this_fn, void *call_site)
  __attribute ((no_instrument_function));
void __cyg_profile_func_exit(void *this_fn, void *call_site)
{
  write(1, "exit\n", 5);
}
#endif

#if DBGS
static char * printzzob(ZeroZeroOneBuf * zb)
/* ; to avoid prototype generation */
{
  static char buf[128];
  sprintf(buf, "{%x, %d -> %d}", mb->data - mb->p.c_buf, mb->p.len, mb->len);
  return buf;
}
#define PMD printzzob(zzob)
#endif

#if TEST
static bool s__fillbuf(ZeroZeroOneBuf * zzob);
static unsigned char * s__memmem(const unsigned char * s, size_t l,
				 const unsigned char * b, size_t bl);
#endif

#include "ghdrs/zzob_priv.h"

static bool s__fillbuf(ZeroZeroOneBuf * zzob)
{
  int l;

  unsigned char * bufend = zzob->p.c_buf + zzob->p.c_buflen;
  unsigned char * rpos = zzob->data + zzob->p.len;

  D(1, ("s__fillbuf(%s)\n", PMD));

  if (rpos - bufend >= 0)
    {
      if (zzob->p.len == zzob->p.c_buflen)
	return false; /* raise BufferFullException */

      if (zzob->p.len)
	memmove(zzob->p.c_buf, zzob->data, zzob->p.len);
      zzob->data = zzob->p.c_buf;
      rpos = zzob->data + zzob->p.len;
    }
  int rlen = (bufend - rpos);
  if (rlen > zzob->p.maxread) {
    if (zzob->p.maxread == 0)
      return false;  /* raise EOFException */
    rlen = zzob->p.maxread;
  }

  l = read(zzob->p.c_fd, rpos, rlen /*> 3? 3: rlen */);
  if (l <= 0)
    return false; /* raise EOFException */

  zzob->p.len += l;
  zzob->p.maxread -= l;

  return true;
}

/* Used when \0\0\1 not found in first full buf... */
static bool s__zzob_init__searchmore(ZeroZeroOneBuf * zzob)
{
  if (zzob->p.len == zzob->p.c_buflen)
    {
      int i = zzob->p.len - 2;
      memcpy(zzob->p.c_buf, zzob->p.c_buf + i, 2);
      zzob->data = zzob->p.c_buf;
      zzob->pos += i;
      zzob->p.len = 2;
      return true;
    }
  else
    return false;
}

bool zzob_init(ZeroZeroOneBuf * zzob, int fd,
	       unsigned char * buf, int buflen, off_t maxread)
{
  zzob_set(zzob, fd, buf, buflen, maxread);
  return zzob_reset(zzob);
}

void zzob_set(ZeroZeroOneBuf * zzob, int fd,
	      unsigned char * buf, int buflen, off_t maxread)
{
  zzob->p.c_fd     = fd;
  zzob->p.c_buf    = buf;
  zzob->p.c_buflen = buflen;

  if (maxread < 0) {
      // 1 << 31 || 1 << 63 and then -1
      maxread = 1; maxread <<= 8 * sizeof maxread - 1; maxread--;
  }
  zzob->p.maxread  = maxread;
}

bool zzob_reset(ZeroZeroOneBuf * zzob)
{
  unsigned char * data;

  zzob->p.len = 0;

  zzob->pos  = 0;
  zzob->data = zzob->p.c_buf;
  zzob->len  = 0;

  D(1, ("zzob_reset(fd=%d, %s)\n", fd, PMD));

  while (true)
    {
      if (s__fillbuf(zzob))
	{
	  data = s__memmem(zzob->data, zzob->p.len,
			   (const unsigned char *)"\000\000\001", 3);
	  if (data != NULL)
	    {
	      zzob->pos += data - zzob->data;
	      zzob->p.len -= data - zzob->data;
	      zzob->data = data;
	      return true;
	    }
	}
      else
	if (!s__zzob_init__searchmore(zzob))
	  return false;
    }
}

bool zzob_data(ZeroZeroOneBuf * zzob, bool locked)
{
  unsigned char * p;
  int o;

  D(1, ("zzob_data(%s)\n", PMD));

  if (locked == false)
    {
      zzob->p.len -= zzob->len;
      zzob->data += zzob->len;
      zzob->pos += zzob->len;
      zzob->len = 0;
    }

  /* if (locked) outf("locked!\n"); */

  o = zzob->len + 3;

  while (true)
    {
      if (zzob->p.len - o > 0
	  && (p = s__memmem(zzob->data + o, zzob->p.len - o,
			    (const unsigned char *)"\000\000\001", 3)) != NULL)
	  break;

      if (! s__fillbuf(zzob))
      {
	D(1,
	  ("s__fillbuf(zzob, zzob->p.len) returned false (%s).\n", PMD));
	if (locked || zzob->p.len < 4)
	  return false;

	p = zzob->data + zzob->p.len;
	break;
      }
    }

  zzob->len = p - zzob->data;

  D(1, ("zzob_data end: %s\n", PMD));
  return true;
}

int zzob_rest(ZeroZeroOneBuf * zzob, bool lastconsumed)
{
  if (lastconsumed)
    {
      zzob->data += zzob->len;
      zzob->p.len -= zzob->len;
      zzob->pos += zzob->len;
    }
  zzob->len = zzob->p.len;
  return zzob->len;
}

/* --- util funcs --- */

static unsigned char * s__memmem(const unsigned char * s, size_t l,
				 const unsigned char * b, size_t bl)
{
  const unsigned char * ss = s;

  D(1, ("s__memmem(l = %d)\n", l));

  while ((s = memchr(ss, b[0], l)) != NULL)
    {
      l -= (s - ss);
      D(0, ("s - ss = %d, l = %d\n", s - ss, l));

      if (l < bl)
	return NULL;

      if (memcmp(s, b, bl) == 0)
	{
	  D(1, ("s\n"));
	  return (unsigned char *)s; /* unavoidable compiler warnings !!! */
	}
      ss = s + 1;
      l--;
    }
  return NULL;
}
